var/spack/repos/builtin/packages/rocm-opencl/package.py

```diff
@@ -64,3 +65,12 @@ def cmake_args(self):
             (self.spec['hip-rocclr'].prefix)
         ]
         return args
+
+    @run_after('install')
+    def post_install(self):
+        vendor_config_path = '/etc/OpenCL/vendors'
```

Comment:
I don't know much about opencl, but this looks very dangerous to me. It requires root permissions too (?). Is it really necessary to do this?

Reply:
without this file, opencl applications cant find GPU "ERROR: clGetPlatformIDs(-1001)" & we will get issue https://github.com/spack/spack/issues/18073.

Reply:
@adamjstewart how do we deal with packages that actually require system changes?

Reply:
We try to avoid them if at all possible. Is there any way to do this in the installation prefix? What happens if this file already exists? What happens if you have multiple versions installed? I think the way Homebrew handles this kind of thing is to print a caveat saying you need to modify this yourself in order to get it to work.

Reply:
OpenCL applications are always expected to link with libOpenCL.so and that internally does a dlopen of the vendor specific implementation, in this case libamdocl.so . ICD loader for OpenCl always expect this path to be in "/etc/OpenCL/vendors". ROCm currently uses the public open source implementation of ICD loader and unless we change the code I do not think we have a different way to map this.

To manage multiple version install and removal we have version specific copy of the file to avoid accidental removal of this files from "/etc/OpenCL/vendors" folder when a user remove one version of ROCm. 
Path resolution of OpenCL.so -> amdocl.so is done using the RUNPATH set into OpenCL.so.

Hope this clarifies.

Reply:
Thanks for the clarification! I wonder if you could convince ICD loader to support `LD_LIBRARY_PATH` like every other application, or RPATH would be even better as that's what Spack uses by default.

Reply:
I do not think AMD maintain this ICD loader it comes from Khronos I believe. This need to be checked if such a change in the code is ok to absorb internally and maintain. Path resolution of OpenCL.so -> amdocl.so is already taken care by LD_LIBRARY_PATH/RUNPATH as long as it knows the name of the file to open(amdocl.so).  Question is how icd loader find the name of the vendor specific library. Directly encoding the library name inside the icdloader code will break the portability of OpenCL to any other vendors. This need some code level analysis to change it without breaking the portability.

Reply:
we have env variable OCL_ICD_VENDORS to control .icd file. https://github.com/RadeonOpenCompute/ROCm-OpenCL-Runtime/tree/master-next/khronos/icd.
@adamjstewart do we have any way to handle this in Spack itself?

Reply:
Yes, you can use `setup_build_env` to set this env var at build-time and `setup_run_env` to add it to the module file at run-time.

Reply:
@adamjstewart, yes it helps. Please review the changes.